ENTMQSTPR-14: Examples for Proxy / Record Encryption (#449)
jboss-container-images/amqstreams-1-openshift-image: examples/proxy/envelope-encryption/PREPARE_KMS.md

## Using an Existing Vault Instance

### Prerequistes

You must have an HashCorp Vault instance running on premise or you have a HashiCorp HCP Vault or Enterprise.  Free tier HCP Vault Secrets is
*not* adequate.

* You must have adminstrative access to the Vault instance or Vault Namespace.
* You must know the Address of the Vault Server, and if Vault Enterprise, the name of the Vault Namespace(s).
* Vault CLI installed.

Note: these examples are prepared against HashiCorp Vault 1.15.

### Configure the Existing Instance

1. Set the `VAULT_ADDR` (and `VAULT_NAMESPACE` if Enterprise) to refer to the Vault Environment.
   ```sh
   export VAULT_ADDR=https://<vault server>:8200
   export VAULT_NAMESPACE=<namespace(s)>
2. Login to Vault as the Adminstrator.
   ```sh
   vault login
   ```
3. Enable the Transit Secrets Engine.
   ```sh
   vault secrets enable transit
   ```
   If the Transit engine was already enabled, an error will be printed but this can be ignored.
4. Create a variable containing the Vault Transit Url.  This will be used later on.
   ```sh
      export VAULT_TRANSIT_URL=${VAULT_ADDR}/v1/${VAULT_NAMESPACE}/transit
   ```

> 5. Update the proxy config to refer to your Vault instance:

Author: FYI If we get https://github.com/kroxylicious/kroxylicious/pull/964 done, these `sed` steps will go away.  Instead, we'll create a config map here.  That CM will be mounted into the pod, and an `envFrom` to pull the CM value into the pod's environment.  Then the interpolation will put the value into the file.

# Preparing a Key Management Service (KMS)

As discussed [here](./README.md) you must provide a KMS solution in order to use the Envelope Encryption Filter.
For Tech Preview, only HashiCorp Vault is supported.

## Preparing HashiCorp Vault

> You may either use an existing instance of HashiCorp Vault you have available in your organisation, or deploy a development instance

Reviewer: ```suggestion
You may either use an existing instance of HashiCorp Vault you have available in your organisation, or deploy a development instance.
```

Reviewer: Can't we use a production instance?

Author: Technically speaking, any Vault instance with the transit engine enable will function just fine.  Should the customer be using their production Vault instance to experiment with a Tech Preview feature? Probably no.   Do we need to spell that out?

> HashoCorp Vault to your OpenShift Cluster which is good enough to try out the Filter.

Reviewer: ```suggestion
Deploying HashiCorp Vault to your OpenShift cluster provides an adequate environment to try out the filter.
```

## Using an Existing Vault Instance

### Prerequistes

> You must have an HashCorp Vault instance running on premise or you have a HashiCorp HCP Vault or Enterprise.  Free tier HCP Vault Secrets is
> *not* adequate.
> 

Reviewer: ```suggestion
You must have an HashCorp Vault instance running on-premise, or access to HashiCorp HCP Vault or Enterprise.  
The free tier of HCP Vault Secrets is *not*  sufficient.

```

> * You must have adminstrative access to the Vault instance or Vault Namespace.
> * You must know the Address of the Vault Server, and if Vault Enterprise, the name of the Vault Namespace(s).
> * Vault CLI installed.
> * GNU sed

Reviewer: ```suggestion
* You must have administrative access to the Vault instance or Vault namespace.
* You must know the address of the Vault server, and if Vault Enterprise, the name of the Vault namespace(s).
* Vault CLI is installed.
* GNU sed is available.
```

> Note: these examples are prepared against HashiCorp Vault 1.15.

Reviewer: ```suggestion
NOTE: These examples are prepared against HashiCorp Vault 1.15.
```

### Configure the Existing Instance

1. Set the `VAULT_ADDR` (and `VAULT_NAMESPACE` if Enterprise) to refer to the Vault Environment.
   ```sh
   export VAULT_ADDR=https://<vault server>:8200
   export VAULT_NAMESPACE=<namespace(s)>
2. Login to Vault as the Adminstrator.
   ```sh
   vault login
   ```
3. Enable the Transit Secrets Engine.
   ```sh
   vault secrets enable transit
   ```
   If the Transit engine was already enabled, an error will be printed but this can be ignored.
4. Create a variable containing the Vault Transit Url.  This will be used later on.
   ```sh
      export VAULT_TRANSIT_URL=${VAULT_ADDR}/v1/${VAULT_NAMESPACE}/transit
   ```
5. Update the proxy config to refer to your Vault instance:
   ```sh
      sed -i "s/\(vaultTransitEngineUrl:\).*$/\1 ${VAULT_TRANSIT_URL}/" */proxy/proxy-config.yaml
   ```  

> 6. To conclude follow the steps below to [Create the Vault Policy and Filter Vault Token](#create-the-vault-policy-and-filter-vault-token)

Reviewer: ```suggestion
6. To complete the process, follow the steps to [Create the Vault Policy and Filter Vault Token](#create-the-vault-policy-and-filter-vault-token)
```

## Deploying a standalone development instance of Vault

> This option will give you an development instance of HashiCorp Vault with ephemeral storage.  This is sufficient to help you evaluate the
> Filter, but is *not* suitable for production or controlled test environments.

Reviewer: ```suggestion
This option deploys a development instance of HashiCorp Vault with ephemeral storage.  This is sufficient to help you evaluate the
filter, but is *not* suitable for production or controlled test environments.
```

> It is assumed that you'll be deploying the HashiCorp Vault feature to the same OpenShift Cluster where you'll be evaluating AMQ Stream Proxy.

Reviewer: ```suggestion
It is assumed that you'll be deploying the HashiCorp Vault feature to the same OpenShift cluster where you'll be evaluating AMQ Stream Proxy.
```

Reviewer: In Streams the _house style_ is we don't capitalize things like _cluster_ only proper nouns like Openshift. I don't know if we want to go along with that?  I won't continue suggesting in this review.

### Prerequsites

* Administrative access to the OpenShift Cluster being used to evaluate AMQ Stream Proxy
* Helm CLI
* OpenShift CLI (oc)

> ### Deploying HashiCorp for Development

Reviewer: ```suggestion
### Deploying HashiCorp Vault for Development
```

1. Log in to the OpenShift Cluster as the adminstrator
   ```sh
   oc login https://<openshift cluster> --username <cluster admin>
   ```
2. Create a Vault Root Token. Keep a note of the token assigned.
   ```sh
   cat /dev/urandom | LC_ALL=C tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1 > vault.root.token
   export VAULT_TOKEN=$(cat vault.root.token)
   ```
3. Install Vault using Helm
   ```sh
   helm repo add hashicorp https://helm.releases.hashicorp.com

>    helm install vault hashicorp/vault --create-namespace --namespace=vault --version 0.27.0 --values vault/helm-dev-values.yaml --set server.dev.devRootToken=${VAULT_TOKEN} --wait

Author: We should refer to the Helm chart by <major>.<minor> so the customer picks up any bug fixes to the chart itself.

      export VAULT_TRANSIT_URL=${VAULT_ADDR}/v1/transit
   ```   
7. To conclude follow the steps below to [Create the Vault Policy and Filter Vault Token](#create-the-vault-policy-and-filter-vault-token)

## Create the Vault Policy and Filter Vault Token

1. Create a Vault Policy for the Envelope Encryption Filter:
   ```sh
   vault policy write amqstreams_proxy_encryption_filter_policy vault/amqstreams_proxy_encryption_filter_policy.hcl
2. Create a Vault Token which will be used by the Envelope Encryption Filter:
   ```bash
   vault token create -display-name "amqstreams-proxy encryption filter" -policy=amqstreams_proxy_encryption_filter_policy -no-default-policy -orphan -field=token > vault.envelope-encryption.token
   ```
3. Prepare a secret containing the Envelope Encryption Vault Token.
   ```bash
   oc create secret generic proxy-envelope-encryption-vault-token -n proxy --from-literal=envelope-encryption-vault-token.txt=vault.envelope-encryption.token --dry-run=client -o yaml > base/proxy/proxy-envelope-encryption-vault-token-secret.yaml
   ```
   The secret file `envelope-encryption-vault-token-secret.yaml` will be applied to the OpenShift Cluster later.

## Cleaning up

> Once you are finished, if you deployed Vault remove it like so:

Reviewer: ```suggestion
After completing the task, if you have deployed Vault, remove it as follows:
```

# Preparing a Key Management Service (KMS)

As discussed [here](./README.md) you must provide a KMS solution in order to use the Envelope Encryption Filter.
For Tech Preview, only HashiCorp Vault is supported.

## Preparing HashiCorp Vault

> You may either use an existing instance of HashiCorp Vault you have available in your organisation, or deploy a development instance

Author: Need to talk about the HashiCorp version (1.15)

HashoCorp Vault to your OpenShift Cluster which is good enough to try out the Filter.

## Using an Existing Vault Instance

> ### Prereqistes

Author: ```suggestion
### Prerequisites
```
